### DISCUSSION TITLE
LMS Automatic Deployment via Ansible

### DISCUSSION THREAD

**Discussion by marioqxx** [LMS-Community/slimserver#902] [General]
I want to create an Ansible tool for deploying LMS. This ideally not only installs the LMS instance, but also automatically installs Plugins and configures LMS as well as the Plugins. So far the CLI seems to be the best approach to configure. What I however failed to achieve is, installing Plugins.
I do not want to follow this here: https://forums.slimdevices.com/forum/user-forums/logitech-media-server/105068-manually-installing-plugin-zips/page3
because the Plugins are not "natively" installed through the PluginManager. In the Extension Plugin I have found the "appsquery" CLI. This seems to call appsQuery, which in turn calls "_appsQueryCB" which then finally calls findUpdates. So there seems to be a way through the CLI to achieve this.
I have added Logging-Output in my local code to Plugin/Extenions/Plugin.pm and tried to figure out what to enter on the CLI. However I am stuck now, as it seems there needs to be an "arg" parameter and its content needs to be a HASH. The command-Parsing in Request.pm however only converts the CLI-string into an array. Anyone here that knows how this CLI-command works or was supposed to work?

Example: Set logging for Extension-Plugin (and CLI-Plugin) to Debug and enter:

appsquery args:something

This will through an error in the log.

Thank you

**Comment by michaelherger** [comment]
I'd recommend you ask this question in the forums, as there are many more users participating in discussions.

You could look at the web UI, what is being sent when you install a plugin through it.

What's the larger context of your plan? Are you trying to deploy to a larger number of systems?

**Reply by marioqxx** [reply]
Thanks for your Reply @michaelherger! I actually did look at the WEB-traffic first, but that didn't work out - or put it that way, I haven't found the correct dataset to be sent that triggers installation. My bigger plan is to get a backup of my system in the way, that I have a "from scratch" install script, that deploys the entire system with all settings anew from a fresh ubuntu or debian install. I have a not yet published repository, but plan to publish soon, and I would like to have LMS-installation and configuration as well covered.
The "appsquery" function seems to have been started to be implemented, but then it seems to be incomplete (This would explain why it is not in the DOCs). Do you think such developer related discussion is well placed in the forum?

**Reply by michaelherger** [reply]
Why not just really back up necessary files? That would be so much simpler than trying to re-configure from scratch... the latter basically means translating the configuration files into instructions to re-build them. Sounds pretty convoluted and will never save you the time you spend trying to figure things out.

**Reply by marioqxx** [reply]
This is true if this is only of use for me, because my settings do not deviate too much from defaults. But if you have a more complicated server setup, with many settings and interactions between server daemons, then this becomes a different point. For example if SSO shall be used to protect access to the LMS-WEB-page. Also, if it is of use to someone else, then this is good enough to invest some time and share. The difference to a backup is, that all settings are "saved" in the ansible-script, similar to a log-book. In case something with the backup-restore goes wrong, it would still be "restorable".

I did came a bit further, but have not yet figured out how the PluginManager works. The appsquery function is not intended to be called through command-line. But with a slight modification in plugins/Extensions/plugin.pm, sub appsQuery add:
`	if (! $request->getParam('args')) { $request->addParam('args' => {type => 'plugin', details => 1}); }`
prior to `my $args = ... ` and in _appsQueryCB add:
`	my $install = $prefs->get($args->{'type'}) || {};`
`	my $params = $request->{'_params'};`
`	for my $k (keys %$params) {`
`		main::DEBUGLOG && $log->debug("_appsQueryCB install $k");`
`		if ( $params->{ $k } eq "install" ) { $install->{ $k } = 1; }`
`	}`
``
`	my $actions = findUpdates($data->{'results'}, $args->{'current'}, $install, $args->{'details'});`
``

then it partly starts to work when additionally editing state.prefs and extensions.prefs.
The plugin-stuff is the complicated one. The config is straight-forward and easy to handle in ansible. So very easily accessible, readable and changeable. No extra code required. If also the plugin-stuff works, then this is presumably very easily configurable as well.

**Reply by michaelherger** [reply]
Could you please provide a diff? Might be easier to understand.

**Reply by marioqxx** [reply]
Sorry, it took me a bit longer to come back to you. I have forked the repository and added the changes to my repository.
https://github.com/Logitech/slimserver/compare/public/8.4...marioqxx:slimserver:public/8.4

**Reply by michaelherger** [reply]
Can you describe what you did, and what the change is supposed to do?

I think what might be simpler and more useful is if one could provide a configuration file telling LMS what plugins to install upon next start. This would allow you to have them configured automatically. And some people who do custom deployments might pre-install some 3rd party plugins which otherwise aren't part of a default installation. Your change would still require LMS to be up and running before you can tell your script to install the plugins.

**Reply by marioqxx** [reply]
I tried to figure out a way how to make plugins installable (and uninstallable) through some command-line interface for automation purposes. If that would work - as you propose - via file, that is equally fine. I just have not found any information about it and the only information I found in this respect is linked above.
So I went on to analyze the source-code and found the CLI and also the "appsquery"-command, which sounded in the comments to be a way to do exactly this. Unfortunately it turned out, that this command is available through CLI, but is not intended to be invoked through it (see above concerning HASH). Now what my added code does is, to convert the CLI-command-line into a HASH, as if it were called from an internal function. If it is called from an internal function, nothing happenes.

How is my test-setup: I run the server on a VM with all defaults. I copy the *.zip file of the plugin to `/var/lib/squeezeboxserver/cache/DownloadedPlugins` folder. I'll then open a command-shell via `telnet localhost 9090` and enter "`appsquery <plugin>:install`". For example "`appsquery CastBridge:install`". If in the protocol for "plugins.extensions" and "plugins.cli" the log-level is set to Debugging you see that the command starts findUpdates function with proper parameters.
Hope this explanation is sufficient, otherwise please let me know and I'll elaborate further without bothering you with superfluous information for you.

**Reply by michaelherger** [reply]
I'm not saying automated plugin installation was already possible in any way. It would be a nice contribution 😉. I've been playing with the thought myself, as there are plugins which IMHO are a must, but aren't part of LMS itself.

**Reply by marioqxx** [reply]
Thank you @michaelherger, understood and actually never understood it differently.

I have played around with it a bit further and I got something to work without code-changes. This works as follows:
**Install:**
- Copy the *.zip archive of the Plugin to `/var/lib/squeezeboxserver/cache/DownloadedPlugins` folder.
- Open `/var/lib/squeezeboxserver/prefs/plugin/state.prefs` file and add a line with the Plugin-name to be installed and set it to `needs-install`. Add another line prefixed with `_ts_` followed by a random-number taken from the plugin above or below.
- Open `/var/lib/squeezeboxserver/prefs/plugin/extensions.prefs` and add the Plugin-name to plugin-dictionary and set status to 1.
- Restart the logitechmediaserver

**Uninstall:**
- Open `/var/lib/squeezeboxserver/prefs/plugin/state.prefs` file and mark the line with the Plugin-name to be un-installed "needs-uninstall". 
- Restart the logitechmediaserver

**Example:**
Install AirBridge-Plugin.
- Copy "RaopBridge.zip" to `/var/lib/squeezeboxserver/cache/DownloadedPlugins`-folder. Pay attention to the permissions.
- Open `/var/lib/squeezeboxserver/prefs/plugin/state.prefs` file and add a line `RaopBridge: needs-install` and further below down `_ts_RaopBridge: 1692520436`. The origin of this number is unknown, so I have chosen the one a line above.
- Open `/var/lib/squeezeboxserver/prefs/plugin/extensions.prefs` and add
  `plugin:`
  `  RaopBridge: 1`
- Restart the logitechmediaserver

Uninstall AirBridge-Plugin.
- Open `/var/lib/squeezeboxserver/prefs/plugin/state.prefs` file and change the line `RaopBridge: enabled` to and `RaopBridge: needs-uninstall`.
- Restart the logitechmediaserver

May I ask you to please test this with your setup and/or plugin. Thanks.

**Reply by michaelherger** [reply]
That's certainly one way to do it, but it's ugly, as you have to deal with the archive outside the process. 

`_ts_*` are just timestamps. I'm not sure it's needed.

**Reply by marioqxx** [reply]
Thanks for clarifying the `_ts_*` names in the config files. Yes, with this approach, the *.zip archive must be downloaded manually. For my purpose this is just perfect, because LMS does handle most things asynchronous. In my case with a slow internet connection, download takes some siginifcant time and eventually even fails. Having the files downloaded makes automation easy. In ansible it would in principle be possible to initiate download and subsequently trigger the installation procedure. So quite conveninent and fully automated.
How would be your idea of command-line based plugin install and uninstall?

**Reply by michaelherger** [reply]
The use case I have in mind (automatically install some 3rd party plugins at LMS installation time) would probably rather be based on a configuration file. LMS already uses a configuration file to understand which plugin to install at startup. But that only works if the file already has been downloaded ahead of time. This should/could be extended to download the files on startup and install them when ready. 

The plugin installer is quite convoluted, unfortunately. But I'm pretty sure this should be feasible with not many code changes. But to figure out what code changes, that's the complicated part 😑.

**Reply by marioqxx** [reply]
True. The plugin-installer code would need a cleanup. It contains some legacy parts and in general the async-setup makes it quite complicated to analyze. I got stuck with my code-analysis and I would have needed a way to log the call-trace. I.e. to see who calls findupdate function when triggered through WEB-interface. Have your got an idea how to accomplish this, I could then dig a bit deeper.

I do have a first working ansible-script that installs may setup and makes some initial configuration. The settings are yaml-style and look like:

    lms_server: logitechmediaserver_8.4.0~1694167045_amd64.deb
    lms_plugins:
      - { name: CastBridge, file: CastBridge-2.2.8.zip }
      - { title: "AirPlay bridge", file: RaopBridge-1.3.3.zip, addfile: raopbridge.xml }
      - { name: UPnPBridge, file: UPnPBridge-2.2.3.zip, addfile: upnpbridge.xml }
      - { name: Spotty, file: Spotty-4.8.8.zip }
      - { name: MaterialSkin, file: lms-material-3.4.4.zip }
    lms_settings:
      - [ "pref", "skin", "material" ]
      - [ "pref", "wizardDone", "1" ]
      - [ "pref", "plugin.castbridge:bin", "squeeze2cast-linux-x86_64-static" ]
      - [ "pref", "plugin.raopbridge:bin", "squeeze2raop-linux-x86_64-static" ]
      - [ "pref", "plugin.upnpbridge:bin", "squeeze2upnp-linux-x86_64-static" ]


This script does automatic download or if "file" is defined copies the file. I'll make this repository public, once I have it complete.

**Reply by marioqxx** [reply]
@michaelherger: I just got one idea, which might work. I did implement the ansible script the complicated way and manipulate the files state.prefs and extensions.prefs manually, i.e. the server must not be running. However it shall be possible to apply the very same procedure I use to configure. Because telnet did not work I settled to the json-API interface.
I tried:
`
curl -g -X POST -d '{"id":1,"method":"slim.request","params":["",["pref","plugin.state:AATest","needs-install"]]}' http://myhost:9000/jsonrpc.js
`
It did what it is supposed to do. It did add `AATest` and the associated `_ts_AATest` lines to `state.prefs`. If you replace AATest with the plugin-name, the first step of my above described procedure is completed. Note: the `_ts_`-part is needed, without it my tests failed.
Manipulating the `extensions.prefs`-file works in the same way:
Reading works with:
`
curl -g -X POST -d '{"id":1,"method":"slim.request","params":["",["pref","plugin.extensions:plugin","?"]]}' http://myhost:9000/jsonrpc.js
`
writing with:
`
curl -g -X POST -d '{"id":1,"method":"slim.request","params":["",["pref","plugin.extensions:plugin",{"HaHa":1,"HoHo":1}]]}' http://myhost:9000/jsonrpc.js
`
because the plugin-setting is a dictionary. If it were a list, such as mediadirs, then the "{}" needs to be replaced by "[]":
`
curl -g -X POST -d '{"id":1,"method":"slim.request","params":["",["pref","mediadirs",["/media","/opt"]]]}' http://myhost:9000/jsonrpc.js
`
Concerning your comment:
"LMS already uses a configuration file to understand which plugin to install at startup. But that only works if the file already has been downloaded ahead of time."
Is there an easier way than the one above?

**Reply by michaelherger** [reply]
I don't think there's any good solution to automatically install 3rd party plugins yet. That's why I hoped you'd come up with a nice pull request 😁.

**Reply by marioqxx** [reply]
I have created the repository and a quick README to give you an idea how this looks like. You will find it here: https://github.com/marioqxx/lms/tree/main.
I certainly understand your desire, having some functionality added to the code. I do not really have an idea what the use-case would be and how this should look like. You stated that you want to ship some Plugins (and install them) with the *.deb package. This shall be easily accomplished with my above explanations. In the *.deb package place the archives in the package such, that after installation they are in the DownloadedPlugins-directory and change the extensions.prefs and state.prefs files.
If you want a separate config-file, then you need additional code that parses the config-file and run the commands. This once again could be accomplished with a short script and the above curl-commands. In this script, there can be easily some lines of code added for automated plugin-download.

On the other hand, the Download in LMS is async and triggers re-start upon completion of download. Any script would be more complicated to monitor download progress from the LMS than any own written code (which is basically calling wget).

So can you please specify what exactly you look for.

**Reply by marioqxx** [reply]
@michaelherger: I think I have found the anchor point where the WEB-page hooks in. It is `Slim/Plugin/Extensions/Settings.pm`. Within that file `sub handler` is called. In line 74 of that file you will find: "# set policy for which plugins are installed/uninstalled etc". You also see the complication I mentioned above regarding async download at line 146 and line 271 (You need to wait until download is complete and Restart-server message is shown to complete Plugin-installation from WEB-GUI). If the server is restarted during download, then everything is back to beginning and requested plugin-installation is forgotten it seems.

The sub handler in turn seems to be linked somehow to `Slim/Plugin/Extensions/Settings.pm`, sub `initPlugin`, line 154.

**Reply by michaelherger** [reply]
That's the handler for the web page where you can install plugins. In order to trigger installation from outside the server itself (using a file, at startup), that's probably an entry point to take inspiration from. But it won't be called unless you hit the running server using a web request. Which is not exactly what we're looking for, is it?
